clowder-framework/clowder
# Update Swagger documentation

## Description

### Problem
Endpoints missing in swagger, some items have incorrect paths, missing responses. Furthermore, the order of the endpoint in Swagger can make it difficult to understand how to use the API

Does not fully address #151, but should cut down a good chunk of the work there

### Approach
* Reorder the endpoint within their categories in Swagger so that `GET/POST /resource` tends to come first, followed by `GET/PUT/DELETE /resource/{id}`
* Fix incorrect parameters notation (use `{param}` instead of `:param`)
* Ensure that path/query parameters are properly represented
* Add (some) missing return codes (e.g. `403` for admin endpoints, `404` for `NotFound`)
* Add many of the missing endpoints with basic info / parameters / return codes
    * Under `/files/{id}`:
       * GET `/paths`
       * POST `/sendArchiveRequest` + `/archive` + `/sendUnarchiveRequest` + `/unarchive`
    * Under `/datasets/{id}`:
      * GET `/downloadFolder` + `/downloadPartial`
      * POST `/queueArchival` + `/queueUnarchival`
    * DELETE `/files/{id}/extractions/{msg_id}` + `/datasets/{ds_id}/extractions/{msg_id}`
    * POST `/previews` + `/reindex`
    * GET `/me` + `/status` + `/tree/getChildrenOfNode` + `/3dTextures`
    * GET `/selected` + all sub-endpoints
    * GET/POST/PUT/DELETE `/proxy` + all sub-endpoints
    * GET `/queries/{id}`

### How to Test
1. Checkout and run this branch locally
2. Navigate to https://clowderframework.org/swagger/?url=http://localhost:9000/swagger
    * You should see Swagger UI appear (NOTE: this is reading from http://localhost:9000/swagger)
3. At the top-left (under "Servers") make sure that `http://localhost:9000/api` is selected
4. At the top-right, click Authorize to authenticate your request
   * You should be able to use username/password or specify an API key to send with all requests
5. After Authorizing, expand one or more or the "tags" listed here
    * You should see that many endpoints have been reordered to put the more commonly-used endpoints near the top of each "tag"
    * For example: "datasets" should start by listing POST+GET `/datasets`, followed by GET+DELETE `/datasets/{id}`, then finally listing the other less-common endpoints
6. Click "Try It Out" beside various endpoints to test their execution and ensure that their behavior is properly documented
    * You should see that many of the endpoints now list 200 and 404 as possible responses 
    * You should see that the admin endpoints also list 403 as a possible response
    * NOTE: Response objects may still be incorrect, but this should be fixed by a later part of #151 

NOTE: If you see "Not Authorized" at any point, please revisit Steps 3 and 4 to ensure that you are still accessing the correct server with the correct credentials

## Review Time Estimate


## Types of changes


## Checklist:


- [x] My change ~~requires~~ is a change to the documentation.
